### PR TITLE
ENG-15836: Wrong answer with "IN", and "<=" when index is present

### PR DESCRIPTION
**Pull request by Chaozz**
The root cause is a bug in the `nestloopindexexecuter`,  literally the `initial_expression`.
### what is `initial_expression`
We extract all the predicate exprs that have the indexed column and concatenate them with **AND** to build the `initial_expression`.
### when `initial_expression` is used?
When we doing an index scan with a LTE(less than or equal to) operator, we find the entry whose key is greater than search key, then do a forward scan using `initial_expression`(we evaluate the `initial_expression` for each input tuple, move forward until it returns false, indicate that we just pass the first failed entry, then we move 1 step backward) to find the correct start point to do a reverse scan.
### the bug
```sql
CREATE TABLE T1 (
    a integer NOT NULL,
    b INTEGER,
 );

 CREATE INDEX T1_IDX ON T1 (a, b);

INSERT INTO T1 VALUES( -1, 9);
INSERT INTO T1 VALUES( 0, 1);
INSERT INTO T1 VALUES( 0, 2);
INSERT INTO T1 VALUES( 1, 3);

SELECT * FROM T1 WHERE a IN (-1, 0, 1) AND b <= 100;
```
the query will be planed as an index nested loop join of table T1 and the in-list(-1,0,1).
```
RETURN RESULTS TO STORED PROCEDURE
 NESTLOOP INDEX INNER JOIN
  inline INDEX SCAN of "T1" using "T1_IDX"
  reverse range-scan covering from (A = column#0) AND (B <= ?0) while (inner-table.column#0 = column#0), filter by (NOT (inner-table.column#1 IS NULL))
  MATERIALIZED SCAN of SQL-IN-LIST (Sort INVALID)
```
and the `inital_expression` is
```
+ Expression[CONJUNCTION_AND, 20]
   ConjunctionExpression
   left:  
   + Expression[COMPARE_LESSTHANOREQUALTO, 14]
      ComparisonExpression
      left:  
      + Expression[VALUE_TUPLE, 32]
         Optimized Column Reference[1, 1]
      right: 
      + Expression[VALUE_PARAMETER, 31]
         OptimizedParameter[0]
   right: 
   + Expression[COMPARE_IN, 17]
      ComparisonExpression
      left:  
      + Expression[VALUE_TUPLE, 32]
         Optimized Column Reference[1, 0]
      right: 
      + Expression[VALUE_VECTOR, 35]
         VectorExpression
```
note that the `initial expression` is always comparing the T1.a with **IN** [-1,0,1].
Let's say if the outer table(in-list) is at tuple (-1), and the inner table(T1) is at tuple (0,1); obviously -1!=0,
it is not a match, but the  `initial expression` is comparing T1.a(0) with the list [-1,0,1] (not -1), so it will return true (indicate a good match) which is wrong!
### fix
Rewriting the initial expression is complicated and may hit other edge cases, so I deprecated the `initial expression` with a series of NullAsMaxComparators in EE which treat NULL values(in the search key) as maximum other than the minimum. Then we can find the LTE entry without `initial expression`.

I will write another ticket to remove the initial expression both from the code and the plans after I merged this.

### CODE REVIEW

File: src/ee/common/tabletuple.h [hunk 1]
```diff
@@ -1274,6 +1275,23 @@ inline int TableTuple::compare(const TableTuple &other) const {
     return VALUE_COMPARE_EQUAL;
 }
 
+/**
+ * Compare two tuples. Null value in the rhs tuple will be treated as maximum.
+ */
+inline int TableTuple::compareNullAsMax(const TableTuple &other) const {
+    const int columnCount = m_schema->columnCount();
+    int diff;
```
Comment (kaikaizi) [inline src/ee/common/tabletuple.h]: add assert that `columnCount == other.m_schema->columnCount()`

File: src/ee/common/tabletuple.h [hunk 2]
```diff
@@ -1274,6 +1275,23 @@ inline int TableTuple::compare(const TableTuple &other) const {
...
+    for (int ii = 0; ii < columnCount; ii++) {
+        const NValue lhs = getNValue(ii);
```
Comment (kaikaizi) [inline src/ee/common/tabletuple.h]: Use reference to make it cheaper.

File: src/ee/executors/indexscanexecutor.cpp [hunk 3]
```diff
@@ -467,26 +467,9 @@ bool IndexScanExecutor::p_execute(const NValueArray &params)
...
-                        // just passed the first failed entry, so move 2 backward
-                        tableIndex->moveToBeforePriorEntry(indexCursor);
-                        break;
-                    }
-                }
-                if (tuple.isNullTuple()) {
-                    tableIndex->moveToEnd(false, indexCursor);
-                }
-            }
+            // find the entry whose key is less or equal than search key
```
Comment (kaikaizi) [inline src/ee/executors/indexscanexecutor.cpp]: less than or equal to
or,
<=

File: src/ee/executors/indexscanexecutor.cpp [hunk 4]
```diff
@@ -467,26 +467,9 @@ bool IndexScanExecutor::p_execute(const NValueArray &params)
...
+            // as the start point to do a reverse scan
+            tableIndex->moveToKeyOrLess(&searchKey, indexCursor);
```
Comment (kaikaizi) [inline src/ee/executors/indexscanexecutor.cpp]: maybe `toLowerBoundInclusive` is more intuitive?

File: src/ee/indexes/indexkey.h [hunk 5]
```diff
@@ -611,7 +642,13 @@ struct GenericComparator
     int compareWithoutPointer(const GenericKey<keySize> &lhs, const GenericKey<keySize> &rhs) const {
         return operator()(lhs, rhs);
     }
-private:
+
+    const std::unique_ptr<GenericNullAsMaxComparator<keySize>> getNullAsMaxComparator() const {
```
Comment (kaikaizi) [inline src/ee/indexes/indexkey.h]: why not just return a type `GenericNullAsMaxComparator<keySize>`?

File: src/ee/indexes/indexkey.h [hunk 6]
```diff
@@ -778,7 +846,12 @@ struct TupleKeyComparator
         return operator()(lhs, rhs);
     }
 
-private:
+    const std::unique_ptr<TupleKeyNullAsMaxComparator> getNullAsMaxComparator() const {
```
Comment (kaikaizi) [inline src/ee/indexes/indexkey.h]: ditto

File: src/ee/indexes/indexkey.h [hunk 7]
```diff
@@ -860,6 +952,12 @@ struct ComparatorWithPointer : public KeyType::KeyComparator {
     int compareWithoutPointer(const KeyWithPointer<KeyType> &lhs, const KeyWithPointer<KeyType> &rhs) const {
         return KeyType::KeyComparator::operator()(lhs, rhs);
     }
+
+    const std::unique_ptr<NullAsMaxComparatorWithPointer<KeyType>> getNullAsMaxComparator() const {
```
Comment (kaikaizi) [inline src/ee/indexes/indexkey.h]: ditto

File: src/ee/executors/indexscanexecutor.cpp (hunk 4 above)
Comment (Chaozz) [reply src/ee/executors/indexscanexecutor.cpp]: There is a `moveToKeyOrGreater` before, so I think `moveToKeyOrLess` is more compatible.

File: src/ee/common/tabletuple.h (hunk 2 above)
Comment (Chaozz) [reply src/ee/common/tabletuple.h]: done

File: src/ee/common/tabletuple.h (hunk 1 above)
Comment (Chaozz) [reply src/ee/common/tabletuple.h]: done

File: src/ee/indexes/indexkey.h (hunk 5 above)
Comment (Chaozz) [reply src/ee/indexes/indexkey.h]: done

File: src/ee/executors/indexscanexecutor.cpp (hunk 3 above)
Comment (Chaozz) [reply src/ee/executors/indexscanexecutor.cpp]: done

File: src/ee/common/tabletuple.h (hunk 1 above)
Comment (kaikaizi) [reply src/ee/common/tabletuple.h]: minor: we use `vassert` these days...

**Pull request merged**